extensions/ql-vscode/test/vscode-tests/minimal-workspace/debugger/save-before-start.test.ts

```diff
@@ -0,0 +1,137 @@
+import * as path from "path";
+import {
+  Position,
+  TextDocument,
+  ViewColumn,
+  commands,
+  window,
+  workspace,
+} from "vscode";
+import { saveAllInGroup } from "../../../../src/run-queries-shared";
+
+/**
+ * Returns a new `TextDocument` with the given file name.
+ *
+ * @param file Path to the file to open, or undefined to create an untitled document.
+ * @param isDirty True if the document should have an edit applied.
+ * @param viewColumn The `ViewColumn` in which to open the file's editor.
+ * @returns The new `TextDocument`
+ */
+async function mockDocument(
+  file: string | undefined,
+  isDirty: boolean,
+  viewColumn: ViewColumn,
+): Promise<TextDocument> {
+  let doc: TextDocument;
+  if (file !== undefined) {
+    doc = await workspace.openTextDocument(
+      path.join(workspace.workspaceFolders![0].uri.fsPath, file),
+    );
+    const editor = await window.showTextDocument(doc, viewColumn, true);
+    if (isDirty) {
+      await editor.edit((edit) => {
+        edit.insert(new Position(0, 0), "test");
+      });
+    }
+  } else {
+    doc = await workspace.openTextDocument({
+      content: 'select "untitled"',
+      language: "ql",
+    });
+    await window.showTextDocument(doc, viewColumn, true);
+  }
+
+  return doc;
+}
+
+/**
+ * Returns a promise that resolves after the given number of milliseconds, and returns the string
+ * "timeout".
+ */
+function delay(ms: number): Promise<string> {
+  return new Promise((resolve) => setTimeout(() => resolve("timeout"), ms));
+}
+
+jest.setTimeout(10000); // A little extra time for the save dialog to appear.
+
+const dirtyFile = "debugger/dirty.ql";
+const cleanFile = "debugger/clean.ql";
+const otherGroupDirtyFile = "debugger/other-dirty.ql";
+const otherGroupCleanFile = "debugger/other-clean.ql";
+
+describe("saveBeforeStart", () => {
+  // We can't easily mock `TextDocument` because its properties are read-only and/or not
+  // configurable, so we rely on the actual `save()` method and the actual `isDirty` property.
+
+  beforeEach(async () => {
+    await commands.executeCommand("workbench.action.closeAllEditors");
+  });
+
+  it("should not save untitled documents without `includeUntitled`", async () => {
+    const dirtyDoc = await mockDocument(dirtyFile, true, ViewColumn.One);
+    const cleanDoc = await mockDocument(cleanFile, false, ViewColumn.One);
+
+    await saveAllInGroup(false);
+
+    expect(dirtyDoc.isDirty).toBe(false);
+    expect(cleanDoc.isDirty).toBe(false);
+  });
+
+  it("should not save dirty documents in other tab groups", async () => {
```

Comment:
This feature is surprising to me. I would expect all docs in the workspace to be saved when I invoke the debugger. Is this the behaviour of this setting in the vscode?

Reply:
:facepalm: I was tricked by the name, but the VS Code source code just calls the internal equivalent of `saveAll()`. The actual behavior for other languages saves tabs in other groups as well. I'll fix it.

My bet is that somebody thought that saving only the current tab group was a good idea early on, so that's what the setting was named. Then, when every developer everywhere tried it and hated it, they just made it implicitly mean "save all".

Reply:
I've changed to code to just use `saveAll()`. The handling of untitled documents appears to be slightly different, but close enough that I won't bother emulating the VS Code behavior precisely.

The original test for `saveAllInGroup()` is now irrelevant, so instead I added a debugger test case to make sure the documents get saved at the right points. Since the source files are modified by the test, this required making a temporary copy of the `data` directory to use as the workspace, instead of opening the committed `data` directory directly. I had to move the test queries _into_ the data directory that we use as the workspace folder, too, since we were pulling them directly from source.